### PR TITLE
Add type hints for Whisper models

### PR DESCRIPTION
**Pull request by donelianc**
# What does this PR do?

Adding type hints for ` Whisper` model (TensorFlow). Related to the issue #16059.

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests), Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? _Task requested [here](https://github.com/huggingface/transformers/issues/16059#issuecomment-1324302192)._
- [x] Did you make sure to update the documentation with your changes? Here are the [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests? _Ran `make fixup` before last commit._

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by donelianc** [comment]
Hi @Rocketknight1, can you help me understand failed test? The following [message](https://app.circleci.com/pipelines/github/huggingface/transformers/52274/workflows/ccc607ce-79a5-4b31-a567-22e5e638338b/jobs/626467?invite=true#step-111-353) comes from the CI test logs:

```
You have to specify either decoder_input_ids or decoder_inputs_embeds
```

According to Whisper [documentation](https://huggingface.co/docs/transformers/v4.24.0/en/model_doc/whisper#transformers.TFWhisperForConditionalGeneration.call.past_key_values),

> If decoder_input_ids and decoder_inputs_embeds are both unset, decoder_inputs_embeds takes the value of inputs_embeds.


But, when I checked the argument `inputs_embeds` set for testing, I noticed that `inputs_embeds=None`.
I'm confused because the model had `=None` for all its parameters before I added the type hints.

**Comment by ArthurZucker** [comment]
Hey! I think you are right here, the documentation is misleading. This only happens for the `...ForConditionalGeneration` and not for the model. Opening a PR right now to fix this

File: src/transformers/models/whisper/modeling_tf_whisper.py [hunk 1]
```diff
@@ -1244,23 +1251,23 @@ def resize_token_embeddings(self, new_num_tokens: int) -> tf.keras.layers.Embedd
...
+        cross_attn_head_mask: Optional[Union[np.ndarray, tf.Tensor]] = None,
+        encoder_outputs: Optional[Tuple[Tuple[Union[np.ndarray, tf.Tensor]]]] = None,
+        past_key_values: Optional[Tuple[Tuple[Union[np.ndarray, tf.Tensor]]]] = None,
+        decoder_inputs_embeds: Optional[Tuple[Union[np.ndarray, tf.Tensor]]] = None,
+        labels: Optional[Union[np.ndarray, tf.Tensor]] = None,
+        use_cache: Optional[bool] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        training: bool = False,
```
Comment (ArthurZucker) [inline src/transformers/models/whisper/modeling_tf_whisper.py:1269]: While here, it should be okay as the inputs are shifted to produce the `decoder_input_ids` if both are not provided

**Comment by donelianc** [comment]
@Rocketknight1 this PR is ready for review. 

Besides including the type hints for main Whisper models, I changed the `output_type` to  `TFSeq2SeqModelOutput` in the `TFWhisperModel`'s _docstrings_ since it has a wrong output (probably a confusion with the `TFWhisperForConditionalGeneration` model).

**Review by Rocketknight1 — APPROVED** [review]
This looks great, thanks! And sorry for missing the earlier ping - I'm not sure why that slipped by me. The docstring fix looks good too, and is definitely appreciated.

**Pull request merged**